### PR TITLE
refactor(web): reworks VisualKeyboard layout spec design 📐

### PR DESCRIPTION
**Pull request by jahorton**
This PR follows #5451 and aims to redesign the parameters needed for how `VisualKeyboard` handles its internal layout calculations.  At present, the various layout objects look up these properties from the `OSKManager` in a manner quite tied to separate 'desktop' and 'touch' tracks, which themselves aren't that well documented.  The critical aspect for layout logic being 'desktop' vs 'touch' has also conflated a lot of the reasoning behind _why_ their separate tracks make the layout calculations that they do.

## User Testing

_**All user testing for this PR has been delegated to #5462, this PR's followup.**_

The exact same stuff will be affected there too, with even _more_ changes in #5462.  It'll be far more efficient to test them both simultaneously, especially given our user-testing bottleneck; the test suite is pretty extensive.

Obviously, this means that this PR shouldn't be merged until _both_ are approved and merged.

### CODE REVIEW

File: web/source/osk/oskManager.ts [hunk 1]
```diff
@@ -196,10 +196,7 @@ namespace com.keyman.osk {
         }
         // For other devices, adjust the object heights, allowing for viewport scaling
       } else {
-        this.vkbd.refreshLayout(this.getKeyboardHeight());
-
-        var b: HTMLElement = this._Box, bs=b.style;
-        bs.height=bs.maxHeight=this.vkbd.computedAdjustedOskHeight(this.getHeight())+'px';
+        this.refreshLayout();
       }
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:200]: Turns out this pattern existed in a few places in a consistent manner - this is how the `OSKManager` has been setting the size of the _touch_ OSK and adjusting its layout.  That operation is part of how the `OSKManager`'s layout itself is managed... thus makes a good starting point for a `OSKManager`-level `refreshLayout`.

File: web/source/osk/oskLayer.ts [hunk 2]
```diff
@@ -85,13 +90,16 @@ namespace com.keyman.osk {
     public refreshLayout(vkbd: VisualKeyboard, paddedHeight: number, trueHeight: number) {
```
Comment (jahorton) [inline web/source/osk/oskLayer.ts:90]: So, fun detail.  Even after the prior PR introduced these objects... turns out this code path was only ever used for touch keyboards.  This PR's changes to these methods allow `refreshLayout` to handle _both_ types, rather than just touch keyboards.

File: web/source/osk/visualKeyboard.ts [hunk 3]
```diff
@@ -172,53 +183,46 @@ namespace com.keyman.osk {
...
+    }
 
-      this.setSize(newWidth, newHeight);
+    /**
+     * Uses fixed scaling for heights of internal elements, rather than relative,
+     * percent-based scaling.
+     */
+    public get usesFixedHeightScaling(): boolean {
+      return this.device.touchable && !this.isStatic;
     }
```
Comment (jahorton) [inline web/source/osk/visualKeyboard.ts:226]: Up 'til now, desktop keyboards and documentation 'keyboards' have always used relative scaling, while touch keyboards have always used fixed (pixel-based) scaling.  Once the full OSK-Core redesign is implemented, these rules won't necessarily hold any longer.

Should someone want a documentation keyboard "rendered" at a fixed size, these will (once fully integrated) allow even a desktop keyboard to be rendered with precise pixel padding and width.  Alternatively, a live touch keyboard could be rendered with percent-based internal layout styling.

File: web/source/osk/visualKeyboard.ts [hunk 4]
```diff
@@ -172,53 +183,46 @@ namespace com.keyman.osk {
...
-        newWidth=0.9*screen.width;
-      }
-  
-      // Default height decision made here: 
-      // https://github.com/keymanapp/keyman/pull/4279#discussion_r560453929
-      newHeight=util.toNumber(c['height'], 0.333 * newWidth);
+    set fontSize(value: ParsedLengthStyle) {
+      this._fontSize = value;
+      this.kbdDiv.style.fontSize = value.styleString;
+    }
```
Comment (jahorton) [inline web/source/osk/visualKeyboard.ts:210]: 1.  We don't automatically `refreshLayout()` here because the `OSKManager` controls when that happens.  Also, because this can be but one call of many that affects the layout; `OSKManager` will delay until all `VisualKeyboard` properties are updated, then do the layout refresh.
2. It was always weird that desktop keyboard font sizes were set here while touch keyboard sizes were set in `OSKManager`.  (See code removed from its `_Load`.)
3. By making it a mutable property on `VisualKeyboard`, it can also be made a mutable property on `OSKManager` in the future, facilitating a decent feature.

File: web/source/osk/visualKeyboard.ts [hunk 5]
```diff
@@ -172,53 +183,46 @@ namespace com.keyman.osk {
...
-      if(this._height) {
-        return this._height;
-      } else {
-        this.loadSizeFromCookie();
-        return this.height;
-      }
+      return this._height;
     }
 
-    protected loadSizeFromCookie() {
```
Comment (jahorton) [inline web/source/osk/visualKeyboard.ts:194]: No more cookies for `VisualKeyboard`; `OSKManager` now fully manages the OSK's width and height, passing updates along to `VisualKeyboard` as needed.

`refreshLayout` is used once those values are available.  Always after `_Load` (particularly, 'after the OSK is inserted into the DOM').

File: web/source/osk/oskManager.ts [hunk 6]
```diff
@@ -766,6 +850,10 @@ namespace com.keyman.osk {
         Ls.visibility='hidden';
       }
 
+      if(this.needsLayout) {
+        this.refreshLayout();
+      }
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:855]: The layout is refreshed early in the process of `_Show`ing it.

File: web/source/osk/oskManager.ts [hunk 7]
```diff
@@ -42,10 +42,37 @@ namespace com.keyman.osk {
...
+     * The computed width for this OSKManager.  May be null if auto sizing
+     * is allowed and the OSKManager is not currently in the DOM hierarchy.
+     */
+    private _computedWidth: number;
+
+    /**
+    * The computed height for this OSKManager.  May be null if auto sizing
+    * is allowed and the OSKManager is not currently in the DOM hierarchy.
+    */
+    private _computedHeight: number;
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:67]: The plain variants will support both relative and absolute specification for an OSK's dimensions:
- percent-based (to fill some portion of the containing element / document)
- pixel-based

In essence, referring to the [OSK-Core design doc](https://docs.google.com/document/d/1uV2IgYv7RQUT318nkJ6Cjwgx8lhGfmtWPqPAD7myfIU/edit#heading=h.cez2bo6bjf6a), this facilitates the different `LayoutHandler` types.
- "static" would use fixed dimensions, while...
- "fill" would use %-based dimensions.

The `_computed` variants will then be used internally to determine the dimensions that the `VisualKeyboard` (along with any other contained elements, like the banner or the desktop header & footer) should fill.

That said, no setters yet, and the only way to 'set' them is explicitly pixel-based at the moment, making `_width` and `_computedWidth` equivalent for now.  This will change after upcoming work.

File: web/source/osk/oskManager.ts [hunk 8]
```diff
@@ -877,6 +956,36 @@ namespace com.keyman.osk {
...
+      } else if(isInDOM) {
+        this._computedWidth   = parseInt(computedStyle.width, 10);
+        this._computedHeight  = parseInt(computedStyle.height, 10);
+      } else {
+        // Cannot perform layout operations!
+        return;
+      }
+
+      // Step 3:  perform layout operations.
+      this.needsLayout = false;
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:978]: Once we know that we can properly perform layout calculations, the first thing we do is determine the internal layout so that everything inside the OSK can have its proper dimensions for use within their own `layoutRefresh` operations.

// TODO:  don't forget to do a `vkbd.setSize(this.computedWidth, this.computedHeight)` here!  (_After_ line 978.)

File: web/source/osk/oskKey.ts [hunk 9]
```diff
@@ -299,13 +300,28 @@ namespace com.keyman.osk {
       return metrics;
     }
 
-    getIdealFontSize(osk: VisualKeyboard, style: {height?: string, fontFamily?: string, fontSize: string}): string {
-      // Recompute the new width for use in autoscaling calculations below, just in case.
-      let emScale = osk.getKeyEmFontSize();
+    getIdealFontSize(vkbd: VisualKeyboard, style: {height?: string, fontFamily?: string, fontSize: string}): string {
+      let buttonStyle = getComputedStyle(this.btn);
```
Comment (jahorton) [inline web/source/osk/oskKey.ts:304]: End goal:  once we can guarantee that `refreshLayout` will always be called at least once after the corresponding `VisualKeyboard` has been added to the DOM, we should be able to drop all parameters for the method and simply rely on `getComputedStyle`.

That guarantee's currently hardest to fulfill for keyboard documentation, since it's built in a state fully detached from the document.  Nothing a temporary `MutationObserver` or `window.setInterval` couldn't 'fix' down the road.

File: web/source/osk/oskManager.ts [hunk 10]
```diff
@@ -780,15 +894,6 @@ namespace com.keyman.osk {
       // TODO:  Move this into the VisualKeyboard class!
       // The following code will always be executed except for externally created OSK such as EuroLatin
       if(this.vkbd) {
-        // Always adjust screen height if iPhone or iPod, to take account of viewport changes
-        // Do NOT condition upon form-factor; this line prevents a bug with displaying
-        // the predictive-text banner on the initial keyboard load.  (Issue #2907)
-        if(device.touchable && device.OS == 'iOS') {
-          this.vkbd.refreshLayout(this.getKeyboardHeight());
-
-          var b: HTMLElement = this._Box, bs=b.style;
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:789]: Why _call_ layout code if this can be part _of_ the layout code itself?  It certainly _relates_ to layout.

File: web/source/osk/oskManager.ts [hunk 11]
```diff
@@ -877,6 +984,39 @@ namespace com.keyman.osk {
       }
     }
 
+    public refreshLayout(): void {
```
Comment (jahorton) [inline web/source/osk/oskManager.ts:987]: Big, important piece of the new paradigm.

Existing OSK methods:
- `_Load`:  whenever a new keyboard becomes active, this "loads" the correct configuration for the keyboard.
    - In the past, also performed some of the OSK's layout operations.
- `_Show`:  responsible for displaying the keyboard.
    - In the past, also performed some of the OSK's layout operations.

While not 100% complete in this PR, one of the main goals here is to splinter those 'layout operation' parts into _this_ method:  `refreshLayout`.  It'll be much more maintainable when all layout aspects are centrally-defined and modularized.

This is designed largely after a common UI lifecycle paradigm.  As expressed in iOS / Swift:
- `will`/`did` `Load`
- `will`/`did` `Appear` (show)
- `layoutIfNeeded`

Even now, KMW actually requires _rebuilding the `VisualKeyboard` from scratch_ whenever the keyboard is rotated.  That's... not exactly efficient.  This'll get us one step closer to being able to reuse the old layout rather than tossing it outright.  (Main remaining issue:  `_Load` (for touch) and cookies (for desktop) are responsible for setting the `width` and `height` properties.)

File: web/source/osk/oskRow.ts [hunk 12]
```diff
@@ -69,27 +69,31 @@ namespace com.keyman.osk {
 
     public refreshLayout(vkbd: VisualKeyboard, rowHeight: number, bottom: number, pad: number) {
       const rs = this.element.style;
-      if(!vkbd.isStatic) {
-        rs.bottom=bottom+'px';
+      if(vkbd.usesFixedHeightScaling) {
```
Comment (jahorton) [inline web/source/osk/oskRow.ts:72]: As mentioned for oskLayer.ts, this function now handles both desktop and touch layout code, not just touch-only code.  It's one step closer to merging the two layout code paths.

**Comment by jahorton** [comment]
OK... there was an issue for a whlie there with embedded keyboard layout behavior.  Required a few changes to get that up and running near-properly, _especially_ on the Android side.  There... for some reason, the simplest answer was simply to nix all in-KMW references to `devicePixelRatio` stuff.

Known issue:  on the first load of the Android app / keyboard, there may be a thin black bar:

image

This goes away after a single rotation, after which it'll always align perfectly in _both_ orientations.  For some reason, the keyboard page isn't initialized with the right value.  As this implies, I happened to find the fix for #5252 while working on this issue; I'll also split that off into its own PR.

While attempting (and failing) to diagnose the remaining issue, I did some tests and can confirm that we should be able to use `window.innerWidth` / `window.innerHeight` to auto-detect the correct dimensions for layout in a future PR, keeping with the OSK-Core design's `FillLayout`.  So... that end-goal should be able to fully eliminate this sort of issue.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboardJSHandler.java [hunk 13]
```diff
@@ -39,7 +39,6 @@ public int getDefaultBannerHeight() {
   @JavascriptInterface
   public int getKeyboardHeight() {
     int kbHeight = KMManager.getKeyboardHeight(context);
-    kbHeight -= kbHeight % 20;
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboardJSHandler.java:42]: Resolves the final "initial thin black bar."

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java [hunk 14]
```diff
@@ -325,6 +326,9 @@ public void onConfigurationChanged(Configuration newConfig) {
     loadJavascript(KMString.format("setBannerHeight(%d)", bannerHeight));
     loadJavascript(KMString.format("setOskWidth(%d)", newConfig.screenWidthDp));
     loadJavascript(KMString.format("setOskHeight(%d)", oskHeight));
+
+    RelativeLayout.LayoutParams params = KMManager.getKeyboardLayoutParams();
+    this.setLayoutParams(params);
```
Comment (jahorton) [inline android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:331]: Matches #5521.

File: android/KMEA/app/src/main/assets/keyboard.html [hunk 15]
```diff
@@ -90,7 +90,11 @@
     }
 
     function getOskHeight() {
-      return oskHeight;
+      var height = oskHeight;
+      if(keyman.osk.banner._activeType != 'blank') {
+        height = height; // Android's perspective - banner height is wholly separate.
+      }
```
Comment (jahorton) [inline android/KMEA/app/src/main/assets/keyboard.html:96]: I was going to have this match how iOS manages things - it changes the set height based on if the banner's displayed.  But... the Android paradigm is different, so I had to roll it back.

Now, just to revert the change while adding documentation for why it's different.

File: web/source/osk/oskManager.ts [hunk 16]
```diff
@@ -231,20 +249,100 @@ namespace com.keyman.osk {
...
+    /**
+     * The computed width for this VisualKeyboard.  May be null if auto sizing
+     * is allowed and the VisualKeyboard is not currently in the DOM hierarchy.
+     */
+    get computedWidth(): number {
+      // if(!this.kbdDiv) {
+      //   // Intermediate state - can be called during VisualKeyboard's constructor, before
+      //   // computedHeight can receive a value.
+      //   return undefined;
+      // }
```
Comment (mcdurdin) [inline web/source/osk/oskManager.ts:277]: Why the commented code?

File: web/source/osk/oskManager.ts [hunk 17]
```diff
@@ -231,20 +249,100 @@ namespace com.keyman.osk {
...
+    /**
+     * The computed height for this VisualKeyboard.  May be null if auto sizing
+     * is allowed and the VisualKeyboard is not currently in the DOM hierarchy.
+     */
+    get computedHeight(): number {
+      // if(!this.kbdDiv) {
+      //   // Intermediate state - can be called during VisualKeyboard's constructor, before
+      //   // computedHeight can receive a value.
+      //   return undefined;
+      // }
```
Comment (mcdurdin) [inline web/source/osk/oskManager.ts:295]: Why the commented code?

File: web/source/osk/oskManager.ts [hunk 18]
```diff
@@ -231,20 +249,100 @@ namespace com.keyman.osk {
...
+      }
+      return this._computedHeight;
+    }
+
+    /**
+     * The top-level style string for the font size used by the predictive banner 
+     * and the primary keyboard visualization elements.
+     */
+    get baseFontSize(): string {
+      return this.parsedBaseFontSize.styleString;
```
Comment (mcdurdin) [inline web/source/osk/oskManager.ts:309]: Is `this.parsedBaseFontSize` always going to be available? It doesn't seem to be set in the constructor.

File: web/source/osk/oskManager.ts [hunk 19]
```diff
@@ -480,13 +578,25 @@ namespace com.keyman.osk {
...
+        const parsedHeight = ParsedLengthStyle.inPixels(height);
+
+        mutatedFlag = mutatedFlag || parsedWidth.styleString  != this._width.styleString;
+        mutatedFlag = mutatedFlag || parsedHeight.styleString != this._height.styleString;
+
+        this._width = parsedWidth;
+        this._height = parsedHeight;
       }
 
+      this.needsLayout = this.needsLayout || mutatedFlag;
```
Comment (mcdurdin) [inline web/source/osk/oskManager.ts:596]:
```suggestion
      this.needsLayout ||= mutatedFlag;
```

Just a thought

File: web/source/osk/oskManager.ts [hunk 20]
```diff
@@ -480,13 +578,25 @@ namespace com.keyman.osk {
...
       if(width && height) {
-        this._baseWidth = width;
-        this._baseHeight = height;
+        mutatedFlag = !this._width || !this._height;
+
+        const parsedWidth = ParsedLengthStyle.inPixels(width);
+        const parsedHeight = ParsedLengthStyle.inPixels(height);
+
+        mutatedFlag = mutatedFlag || parsedWidth.styleString  != this._width.styleString;
+        mutatedFlag = mutatedFlag || parsedHeight.styleString != this._height.styleString;
```
Comment (mcdurdin) [inline web/source/osk/oskManager.ts:590]:
```suggestion
        mutatedFlag ||= parsedWidth.styleString  != this._width.styleString;
        mutatedFlag ||= parsedHeight.styleString != this._height.styleString;
```

File: web/source/osk/visualKeyboard.ts (hunk 5 above)
Comment (mcdurdin) [reply web/source/osk/visualKeyboard.ts:195]: If no more cookies, then what remembers the keyboard size and position between page loads on e.g. keymanweb.com? (not saying that cookies within kmw are the right solution, but we need to make sure that there is some pathway for migration)

File: web/source/osk/visualKeyboard.ts [hunk 21]
```diff
@@ -233,23 +237,28 @@ namespace com.keyman.osk {
...
+        }
+
+        this.refreshLayout();
       }
     }
 
-    public defaultFontSize(): number {
-      return this.height / 8;
+    public defaultDesktopFontSize(): number {
+      return this.height ? this.height / 8 : undefined;
```
Comment (mcdurdin) [inline web/source/osk/visualKeyboard.ts:252]: Should add a doc comment about possible return values

**Review by mcdurdin — CHANGES_REQUESTED** [review]
I like the direction this PR is going. Looking forward to seeing this come in.

I think that there is going to need to be a *lot* of testing for these changes, however, as this PR touches a bunch of assumptions about browsers and layout. We'll need to make sure we test on min supported version of each browser as well as current bleeding edge before release in all the permutations that we support. Sad face time!

File: android/KMEA/app/src/main/assets/keyboard.html (hunk 15 above)
Comment (jahorton) [reply android/KMEA/app/src/main/assets/keyboard.html:96]: Oh, of course.  That note was a reminder-to-self bit; these comments were end-of-day.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java (hunk 14 above)
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:331]: > "is in both"?

Like... both the in-app and the system keyboard?

The reason I thought to put it here:  this is the function that's called whenever the keyboard is rotated.  So, that's the obvious time to actually do code that handles rotation things.  I actually don't know why the black bar wasn't happening in the Android app's system keyboard before this fix, but if I did know, I'd probably knock out the other block of code to confirm that this is sufficient to handle both contexts.  DRY > WET and all that.

If it turns out that this needs tweaks based on the review of #5521, I'll make sure those adjustments are made here too.

File: web/source/kmwembedded.ts [hunk 22]
```diff
@@ -257,11 +257,9 @@ namespace com.keyman.text {
   keymanweb['correctOSKTextSize']=function() {
     let osk: com.keyman.osk.OSKManager = keymanweb.osk;
     if(osk?.vkbd) {
-      osk.vkbd.refreshLayout(osk.getKeyboardHeight());
-
-      var b: HTMLElement = osk._Box, bs=b.style;
-      bs.height=bs.maxHeight=osk.vkbd.computedAdjustedOskHeight(osk.getHeight())+'px';
+      osk.refreshLayout();
 
+      // ... that `refreshLayout` is probably redundant:  _Load should fully reload the keyboard.
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/kmwembedded.ts:262]: Right now, it is redundant.  I do hope to make it _not_ redundant though, but we're not ready to drop the `_Load` part of this call quite yet.

File: web/source/osk/bannerManager.ts [hunk 23]
```diff
@@ -72,10 +72,6 @@ namespace com.keyman.osk {
       // any 'manually set' options come post-construction.
       // This will also automatically set the default banner in place.
       this.setOptions(BannerManager.DEFAULT_OPTIONS);
-
-      // Register a listener for model change events so that we can hot-swap the banner as needed.
-      let keyman = com.keyman.singleton;
-      keyman.core.languageProcessor.on('statechange', this.selectBanner.bind(this));
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/bannerManager.ts:78]: It is; it's just been moved to `OSKManager`, around line 133.

`OSKManager` needs to know about banner state changes in order to call `refreshLayout` when the banner mode changes, thus the overall OSK's height changes.  Its handler includes a call to the original handler, loosely speaking.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java (hunk 14 above)
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:331]: Went and double-checked; the system keyboard works fine.

File: web/source/osk/bannerManager.ts [hunk 24]
```diff
@@ -213,7 +209,7 @@ namespace com.keyman.osk {
      * allowing logic to automatically hot-swap `Banner`s as needed.
      * @param state 
      */
-    private selectBanner(state: text.prediction.StateChangeEnum) {
+    selectBanner(state: text.prediction.StateChangeEnum) {
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/bannerManager.ts:212]: That's not the intent; we're definitely not aiming to document it as such.

Continuing from the prior comment, this is necessary to maintain behavior of the event handler that got moved to `OSKManager`.

File: web/source/osk/oskBaseKey.ts [hunk 25]
```diff
@@ -79,40 +79,46 @@ namespace com.keyman.osk {
...
+        ks.left=this.objectGeometry(vkbd, totalPercent+spec['padpc']);
+        if(!vkbd.isStatic) {
           ks.bottom=rowStyle.bottom;
         }
-        ks.height=rowStyle.height;  // must be specified in px for rest of layout to work correctly
 
-        if(!osk.isStatic) {
+        let parsedRowStyle = new ParsedLengthStyle(rowStyle.height);
+        if(parsedRowStyle.absolute) {
+          ks.height=rowStyle.height;  // must be specified in px for rest of layout to work correctly
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/oskBaseKey.ts:110]: That'll happen in #5462.

File: web/source/osk/oskBaseKey.ts [hunk 26]
```diff
@@ -79,40 +79,46 @@ namespace com.keyman.osk {
...
+        } else {
+          ks.height='100%';
+        }
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/oskBaseKey.ts:113]: We could, but it'd be knocked out immediately by the next PR.

Short answer:  touch-path only b/c inconsistent CSS layout specifications.

The next PR addresses _those_ and makes 'em consistent, which allows simplification here.

File: web/source/osk/oskKey.ts [hunk 27]
```diff
@@ -168,13 +168,7 @@ namespace com.keyman.osk {
...
 
-      // Apply an overriding class for 5-row layouts
-      var nRows=vkbd.kbdLayout['layer'][0]['row'].length;
-      if(nRows > 4 && vkbd.device.formFactor == 'phone') {
-        btn.className='kmw-key kmw-5rows kmw-key-'+OSKKey.BUTTON_CLASSES[n];
-      } else {
-        btn.className='kmw-key kmw-key-'+OSKKey.BUTTON_CLASSES[n];
-      }
+      btn.className='kmw-key kmw-key-'+OSKKey.BUTTON_CLASSES[n];
     }
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/oskKey.ts:172]: Very unlikely.  If my understanding of CSS rules is right, it'd require a precise `.kmw-key.kmw-5rows` selector - one that requires the two classes to be specified directly on the same element in order to match.

In contrast, `.kmw-key .kmw-5rows` would match perfectly fine.  So... a _very_ niche breaking change.

File: web/source/osk/oskKey.ts [hunk 28]
```diff
@@ -424,16 +429,13 @@ namespace com.keyman.osk {
...
+      let specialText = this.renameSpecialKey(keyText, vkbd);
       if(specialText != keyText) {
         // The keyboard wants to use the code for a special glyph defined by the SpecialOSK font.
         keyText = specialText;
         spec['font'] = "SpecialOSK";
       }
 
-      // Grab our default for the key's font and font size.
-      ts.fontSize=osk.fontSize;     //Build 344, KMEW-90
-
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/oskKey.ts:436]: The general `osk.fontSize` is applied at the top level of the `VisualKeyboard`, on its `.kbdDiv` element, which is the layer-group's parent.

Result:  the _default_ font-size is now set for all elements in the same manner as for the desktop OSK.

We still get the value, as it's now factored into the `.getKeyEmFontSize()` method, which is called at line 459.

File: web/source/osk/oskManager.ts (hunk 18 above)
Comment (jahorton) [reply web/source/osk/oskManager.ts:309]: ... that's what the getter that directly follows this block is for.

File: web/source/osk/visualKeyboard.ts (hunk 5 above)
Comment (jahorton) [reply web/source/osk/visualKeyboard.ts:195]: > No more cookies **_for `VisualKeyboard`_** [...]

I never said no more cookies for `OSKManager` - that's where those references belong and remain.

**Comment by jahorton** [comment]
> I think that there is going to need to be a lot of testing for these changes, however

Yep, and it doesn't stop there.  Which is why there's a quite extensive test spec on #5462.  Feel free to suggest extra tests if you have any ideas.

File: web/source/osk/oskKey.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply web/source/osk/oskKey.ts:172]: Still a breaking change but probably not worth worrying about.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java (hunk 14 above)
Comment (mcdurdin) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:331]: actually, meant "is in both PRs" -- this one and #5521!

File: web/source/osk/oskManager.ts (hunk 18 above)
Comment (mcdurdin) [reply web/source/osk/oskManager.ts:309]: Yeah, but could `baseFontSize` ever be called by another component before the getter below is? Just in terms of future robustness.

File: web/source/osk/lengthStyle.ts [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+        return {val: val, absolute: false, special: 'rem'};
+      } else if(spec.indexOf('em') != -1) {
+        val = parseFloat(spec);
+        return {val: val, absolute: false, special: 'em'};
+      } else {
+        // Cannot parse.
+        console.error("Could not properly parse specified length style info: '" + spec + "'.");
+        return null;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (jahorton) [reply web/source/osk/lengthStyle.ts:96]: I do note that we're not checking the assumption of "number in a string without units", but otherwise: 'right'.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java (hunk 14 above)
Comment (jahorton) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboard.java:331]: Oh.  This is where I first made the fix, then I pseudo-🍒'd it to that PR.

Our tester's been labeling tests as failures when the black bar shows up, despite it being a known separate issue.  So, it'll help prevent false-negatives.  Also... the fact is that between this PR and the next, we're tweaking so many related aspects of the OSK that I need to be _sure_ that this PR doesn't turn around and re-break the fix instantly.

File: web/source/osk/lengthStyle.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply web/source/osk/lengthStyle.ts]: Slightly confused by this comment. `parseFloat` accepts a `Number` as input as well as string. So what am I missing?

File: web/source/osk/oskManager.ts (hunk 18 above)
Comment (jahorton) [reply web/source/osk/oskManager.ts:309]: I don't see the issue here.  This _directly calls the getter_, which is written to perform lazy-initialization if/when necessary.  If there's a part of that getter that isn't robust, please point it out to me.

**Comment by jahorton** [comment]
Huh.  So, it turns out that the `||=` operator isn't a thing in TS until 4.0, but KMW currently uses 3.8.3.

File: web/source/osk/oskManager.ts (hunk 18 above)
Comment (mcdurdin) [reply web/source/osk/oskManager.ts:297]: oh duh. I don't know why it didn't click that the `get` was for this...!

**Comment by mcdurdin** [comment]
> Huh. So, it turns out that the `||=` operator isn't a thing in TS until 4.0, but KMW currently uses 3.8.3.

😢

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.96-alpha](https://keyman.com/downloads/releases/alpha/15.0.96)
